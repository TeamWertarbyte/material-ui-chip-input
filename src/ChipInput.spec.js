/* eslint-env jest */
import React from 'react'
import { mount } from 'enzyme'
import Chip from '@material-ui/core/Chip'
import ChipInput from './ChipInput'

describe('uncontrolled mode', () => {
  it('matches the snapshot', () => {
    const tree = mount(
      <ChipInput defaultValue={['foo', 'bar']} />
    )
    expect(tree).toMatchSnapshot()
  })

  it('displays the default value in chips', () => {
    const tree = mount(
      <ChipInput defaultValue={['foo', 'bar', 'foobar']} />
    )
    expect(tree.find('Chip').map((chip) => chip.text())).toEqual(['foo', 'bar', 'foobar'])
  })

  it('displays added chips', () => {
    const tree = mount(
      <ChipInput defaultValue={['foo', 'bar']} />
    )
    tree.find('input').getDOMNode().value = 'test'
    tree.find('input').simulate('keyDown', { keyCode: 13 }) // press enter
    expect(tree.find('Chip').map((chip) => chip.text())).toEqual(['foo', 'bar', 'test'])
  })

  it('calls onChange when adding new chips', () => {
    const handleChange = jest.fn()
    const tree = mount(
      <ChipInput onChange={handleChange} />
    )

    tree.find('input').getDOMNode().value = 'foo'
    tree.find('input').simulate('keyDown', { keyCode: 13 }) // press enter
    expect(handleChange).toBeCalledWith(['foo'])

    tree.find('input').getDOMNode().value = 'bar'
    tree.find('input').simulate('keyDown', { keyCode: 13 }) // press enter
    expect(handleChange).toBeCalledWith(['foo', 'bar'])
  })

  it('calls onChange when deleting chips with backspace key', () => {
    const handleChange = jest.fn()
    const tree = mount(
      <ChipInput defaultValue={['foo', 'bar']} onChange={handleChange} />
    )

    tree.find('input').simulate('keyDown', { keyCode: 8 }) // backspace (to focus the chip)
    tree.find('input').simulate('keyDown', { keyCode: 8 }) // backspace (to remove the chip)
    expect(handleChange).toBeCalledWith(['foo'])
  })

  it('calls onChange when deleting chips with delete key', () => {
    const handleChange = jest.fn()
    const tree = mount(
      <ChipInput defaultValue={['foo', 'bar']} onChange={handleChange} />
    )

    tree.find('input').simulate('keyDown', { keyCode: 8 }) // backspace (to focus the chip)
    tree.find('input').simulate('keyDown', { keyCode: 46 }) // del (to remove the chip)
    expect(handleChange).toBeCalledWith(['foo'])
  })

  it('calls onChange when deleting chips by clicking on the remove button', () => {
    const handleChange = jest.fn()
    const tree = mount(
      <ChipInput defaultValue={['foo', 'bar']} onChange={handleChange} />
    )
    tree.find('Cancel').first().simulate('click')
    expect(handleChange).toBeCalledWith(['bar'])
  })

  it('does not add empty chips', () => {
    const handleChange = jest.fn()
    const tree = mount(
      <ChipInput onChange={handleChange} />
    )

    tree.find('input').getDOMNode().value = ' '
    tree.find('input').simulate('keyDown', { keyCode: 13 }) // press enter
    expect(handleChange).not.toBeCalled()
  })

  it('does not add duplicate chips by default', () => {
    const handleChange = jest.fn()
    const tree = mount(
      <ChipInput defaultValue={['a']} onChange={handleChange} />
    )

    tree.find('input').getDOMNode().value = 'a'
    tree.find('input').simulate('keyDown', { keyCode: 13 }) // press enter
    expect(handleChange).not.toBeCalled()
  })

  it('does add duplicate chips if allowDuplicates is set to true', () => {
    const handleChange = jest.fn()
    const tree = mount(
      <ChipInput defaultValue={['a']} onChange={handleChange} allowDuplicates />
    )

    tree.find('input').getDOMNode().value = 'a'
    tree.find('input').simulate('keyDown', { keyCode: 13 }) // press enter
    expect(handleChange).toBeCalledWith(['a', 'a'])
  })

  it('calls inputRef when it set', () => {
    let inputRef
    mount(
      <ChipInput inputRef={ref => { inputRef = ref }} />
    )

    expect(inputRef.tagName.toLowerCase()).toBe('input')
  })

  it('calls onUpdateInput when the input changes', () => {
    const handleUpdateInput = jest.fn()
    const tree = mount(
      <ChipInput onUpdateInput={handleUpdateInput} />
    )

    tree.find('input').getDOMNode().value = 'foo'
    tree.find('input').simulate('change', { target: tree.find('input').getDOMNode() })
    expect(handleUpdateInput).toBeCalledWith(
     expect.objectContaining({
       target: expect.anything()
     })
    )
  })
})

describe('chip focusing', () => {
  function getFocusedChip (tree) {
    return tree.find('Chip').filterWhere((chip) => chip.getDOMNode().style.backgroundColor !== '')
  }

  function focusChip (tree, name) {
    tree.find('Chip').filterWhere((chip) => chip.text() === name).simulate('click')
  }

  it('focuses a chip on click', () => {
    const tree = mount(
      <ChipInput defaultValue={['foo', 'bar']} />
    )
    tree.find('Chip').at(1).simulate('click')
    expect(getFocusedChip(tree).length).toBe(1)
    expect(getFocusedChip(tree).text()).toBe('bar')
  })

  it('focuses the last chip when pressing backspace', () => {
    const tree = mount(
      <ChipInput defaultValue={['a', 'b', 'c']} />
    )
    tree.find('input').simulate('keyDown', { keyCode: 8 }) // backspace
    expect(getFocusedChip(tree).text()).toBe('c')
  })

  it('focuses the last chip when pressing the left arrow key if the input is empty', () => {
    const tree = mount(
      <ChipInput defaultValue={['a', 'b', 'c']} />
    )
    tree.find('input').simulate('keyDown', { keyCode: 37 }) // arrow left
    expect(getFocusedChip(tree).text()).toBe('c')

    // don't focus the chip if the input is not empty
    tree.find('input').getDOMNode().value = 'd'
    tree.find('input').simulate('keyDown')
    expect(getFocusedChip(tree).length).toBe(0)
    tree.find('input').simulate('keyDown', { keyCode: 37, target: { value: 'd' } }) // arrow left
    expect(getFocusedChip(tree).length).toBe(0)
  })

  it('unfocuses the focused chip while adding a new chip', () => {
    const tree = mount(
      <ChipInput defaultValue={['foo', 'bar']} />
    )
    focusChip(tree, 'foo')
    tree.find('input').simulate('keyDown')
    expect(getFocusedChip(tree).length).toBe(0)
  })

  it('unfocuses the focused chip on blur', () => {
    const tree = mount(
      <ChipInput defaultValue={['foo', 'bar']} />
    )
    focusChip(tree, 'foo')
    tree.find('input').simulate('blur')
    expect(getFocusedChip(tree).length).toBe(0)
  })

  it('unfocuses the focused chip when switching to disabled state', () => {
    const tree = mount(
      <ChipInput defaultValue={['foo', 'bar']} />
    )
    focusChip(tree, 'foo')
    tree.setProps({ disabled: true })
    expect(getFocusedChip(tree).length).toBe(0)
  })

  it('moves the focus to the left when pressing the left arrow key', () => {
    const tree = mount(
      <ChipInput defaultValue={['a', 'b', 'c']} />
    )
    focusChip(tree, 'b')
    tree.find('input').simulate('keyDown', { keyCode: 37 }) // arrow left
    expect(getFocusedChip(tree).text()).toBe('a')

    // keep the first element focused when pressing left if the first element is already focused
    tree.find('input').simulate('keyDown', { keyCode: 37 }) // arrow left
    expect(getFocusedChip(tree).text()).toBe('a')
  })

  it('moves the focus to the right when pressing the right arrow key', () => {
    const tree = mount(
      <ChipInput defaultValue={['a', 'b', 'c']} />
    )
    focusChip(tree, 'b')
    tree.find('input').simulate('keyDown', { keyCode: 39 }) // arrow right
    expect(getFocusedChip(tree).text()).toBe('c')

    // unfocus all chips if the right arrow key is pressed when focusing the last chip
    tree.find('input').simulate('keyDown', { keyCode: 39 }) // arrow right
    expect(getFocusedChip(tree).length).toBe(0)
  })

  it('focuses the chip to the left when removing a chip by pressing backspace', () => {
    const tree = mount(
      <ChipInput defaultValue={['a', 'b', 'c']} />
    )
    focusChip(tree, 'b')

    tree.find('input').simulate('keyDown', { keyCode: 8 }) // backspace
    expect(getFocusedChip(tree).text()).toBe('a')
  })

  it('focuses the chip at the previously focused position when removing a chip by pressing delete', () => {
    const tree = mount(
      <ChipInput defaultValue={['a', 'b', 'c']} />
    )
    focusChip(tree, 'b')

    tree.find('input').simulate('keyDown', { keyCode: 46 }) // delete
    expect(getFocusedChip(tree).text()).toBe('c')
  })
})

describe('placeholder', () => {
  it('displays a placeholder', () => {
    const tree = mount(
      <ChipInput placeholder='Placeholder' />
    )

    expect(tree.find('input').getDOMNode().getAttribute('placeholder')).toBe('Placeholder')
  })

  it('is hidden if there are chips', () => {
    const tree = mount(
      <ChipInput placeholder='Placeholder' value={['foo']} />
    )

    expect(tree.find('input').getDOMNode().getAttribute('placeholder')).toBe(null)
  })

  it('is hidden if there is a floating label', () => {
    const tree = mount(
      <ChipInput placeholder='Placeholder' label='Floating label' />
    )

    expect(tree.find('input').getDOMNode().getAttribute('placeholder')).toBe(null)
  })

  it('is visible if the floating label is floating', () => {
    const tree = mount(
      <ChipInput placeholder='Placeholder' label='Floating label' />
    )
    tree.find('input').simulate('focus')
    expect(tree.find('InputLabel').prop('shrink')).toBe(true)
    expect(tree.find('input').getDOMNode().getAttribute('placeholder')).toBe('Placeholder')
  })
})

describe('floating label', () => {
  it('is displayed', () => {
    const tree = mount(
      <ChipInput label='Floating label' />
    )
    expect(tree.find('InputLabel').text()).toBe('Floating label')
    expect(tree.find('InputLabel').prop('shrink')).toBe(false)
  })

  it('shrinks if there are chips', () => {
    const tree = mount(
      <ChipInput label='Floating label' value={['foo']} />
    )
    expect(tree.find('InputLabel').prop('shrink')).toBe(true)
  })

  it('shrinks if there are no chips but text input', () => {
    const tree = mount(
      <ChipInput label='Floating label' />
    )
    tree.find('input').getDOMNode().value = 'foo'
    tree.find('input').simulate('change')
    expect(tree.find('InputLabel').prop('shrink')).toBe(true)
  })
})

describe('helper text', () => {
  it('is displayed', () => {
    const tree = mount(
      <ChipInput helperText='Helper text' />
    )
    expect(tree.find('FormHelperText').text()).toBe('Helper text')
  })
})

describe('custom chips', () => {
  it('calls a chip renderer for every chip', () => {
    const chipRenderer = jest.fn(({ text }, key) => <Chip key={key} label={text.toUpperCase()} />)
    const tree = mount(
      <ChipInput value={['a', 'b']} chipRenderer={chipRenderer} />
    )
    expect(chipRenderer).toHaveBeenCalledTimes(2)
    expect(tree.find('Chip').map((chip) => chip.text())).toEqual(['A', 'B'])

    expect(chipRenderer.mock.calls[0][0]).toEqual({
      value: 'a',
      text: 'a',
      chip: 'a',
      className: expect.any(String),
      isDisabled: false,
      isFocused: false,
      handleClick: expect.any(Function),
      handleDelete: expect.any(Function)
    })
  })
})

describe('blurBehavior modes', () => {
  it('clears the input on blur', () => {
    const tree = mount(
      <ChipInput value={['a', 'b']} />
      )
    tree.find('input').simulate('change', { target: { value: 'foo' } })
    tree.find('input').simulate('blur')
    expect(tree.find('input').getDOMNode().value).toBe('')
  })

  it('does not clear the input on blur with blurBehavior set to ignore', () => {
    const tree = mount(
      <ChipInput value={['a', 'b']} blurBehavior='ignore' />
      )
    tree.find('input').simulate('change', { target: { value: 'foo' } })
    tree.find('input').simulate('blur')
    expect(tree.find('input').getDOMNode().value).toBe('foo')
  })

  it('adds the input on blur with blurBehavior set to add', () => {
    const handleChange = jest.fn()
    jest.useFakeTimers()
    const tree = mount(
      <ChipInput defaultValue={['a', 'b']} blurBehavior='add' onChange={handleChange} />
      )
    tree.find('input').getDOMNode().value = 'blur'
    tree.find('input').simulate('blur')

    jest.runAllTimers()

    expect(tree.find('input').getDOMNode().value).toBe('')
    expect(setTimeout).toHaveBeenCalledTimes(1)

    setTimeout(_ => {
      expect(tree.find('Chip').map((chip) => chip.text())).toEqual(['a', 'b', 'blur'])
    })
  })
})

describe('keys', () => {
  it('calls onKeyDown prop', () => {
    let keyPressed = null
    const tree = mount(
      <ChipInput value={['a', 'b']}
        onKeyDown={(e) => {
          keyPressed = e.keyCode
        }}
      />
    )
    tree.find('input').simulate('keyDown', {
      keyCode: 40, // down
      metaKey: false,
      ctrlKey: false,
      altKey: false
    })
    expect(keyPressed).toBe(40)
  })

  it('calls onKeyUp prop', () => {
    let keyPressed = null
    const tree = mount(
      <ChipInput value={['a', 'b']}
        onKeyUp={(e) => {
          keyPressed = e.keyCode
        }}
      />
    )
    tree.find('input').simulate('keyUp', {
      keyCode: 123,
      metaKey: false,
      ctrlKey: false,
      altKey: false
    })
    expect(keyPressed).toBe(123)
  })

  it('calls onKeyPress prop', () => {
    let keyPressed = null
    const tree = mount(
      <ChipInput value={['a', 'b']}
        onKeyPress={(e) => {
          keyPressed = e.keyCode
        }}
      />
    )
    tree.find('input').simulate('keyPress', {
      keyCode: 321,
      metaKey: false,
      ctrlKey: false,
      altKey: false
    })
    expect(keyPressed).toBe(321)
  })

  it('stops when onKeyDown prevents the default', () => {
    let preventDefault = false

    function handleKeyDown (e) {
      if (preventDefault) {
        e.preventDefault()
      }
    }

    const tree = mount(
      <ChipInput
        onKeyDown={handleKeyDown}
        newChipKeyCodes={null} /> // This will raise an exception if the code reaches that part
    )
    let input = tree.find('input')

    expect(() => input.simulate('keyDown', {})).toThrow()

    preventDefault = true
    expect(() => input.simulate('keyDown', {})).toBeDefined()
  })

  it('prevents default when chip is added', () => {
    const tree = mount(
      <ChipInput />
    )
    let prevented = false
    tree.find('input').simulate('keyDown', {
      keyCode: 13,
      metaKey: false,
      ctrlKey: false,
      altKey: false,
      target: {value: 'non-empty'},
      preventDefault: () => {
        prevented = true
      }
    })
    expect(prevented).toBe(true)
  })
})
