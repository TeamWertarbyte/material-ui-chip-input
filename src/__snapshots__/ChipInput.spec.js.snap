// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`uncontrolled mode matches the snapshot 1`] = `
<WithStyles(ChipInput)
  defaultValue={
    Array [
      "foo",
      "bar",
    ]
  }
>
  <ChipInput
    allowDuplicates={false}
    blurBehavior="clear"
    classes={
      Object {
        "chip": "ChipInput-chip-14",
        "chipContainer": "ChipInput-chipContainer-4",
        "disabled": "ChipInput-disabled-11",
        "error": "ChipInput-error-13",
        "focused": "ChipInput-focused-10",
        "helperText": "ChipInput-helperText-8",
        "inkbar": "ChipInput-inkbar-9",
        "input": "ChipInput-input-3",
        "inputRoot": "ChipInput-inputRoot-2",
        "label": "ChipInput-label-6",
        "labelShrink": "ChipInput-labelShrink-7",
        "labeled": "ChipInput-labeled-5",
        "root": "ChipInput-root-1",
        "underline": "ChipInput-underline-12",
      }
    }
    clearInputValueOnAdd={false}
    defaultValue={
      Array [
        "foo",
        "bar",
      ]
    }
    inputRef={[Function]}
    newChipKeyCodes={
      Array [
        13,
      ]
    }
  >
    <WithStyles(FormControl)
      className="ChipInput-root-1"
      onClick={[Function]}
    >
      <FormControl
        className="ChipInput-root-1"
        classes={
          Object {
            "fullWidth": "MuiFormControl-fullWidth-18",
            "marginDense": "MuiFormControl-marginDense-17",
            "marginNormal": "MuiFormControl-marginNormal-16",
            "root": "MuiFormControl-root-15",
          }
        }
        component="div"
        disabled={false}
        error={false}
        fullWidth={false}
        margin="none"
        onClick={[Function]}
        required={false}
      >
        <div
          className="MuiFormControl-root-15 ChipInput-root-1"
          onBlur={[Function]}
          onClick={[Function]}
          onFocus={[Function]}
          required={false}
        >
          <div
            className="ChipInput-chipContainer-4 ChipInput-inkbar-9 ChipInput-underline-12"
          >
            <WithStyles(Chip)
              className="ChipInput-chip-14"
              key="0"
              label="foo"
              onClick={[Function]}
              onDelete={[Function]}
              style={
                Object {
                  "backgroundColor": undefined,
                  "pointerEvents": undefined,
                }
              }
            >
              <Chip
                className="ChipInput-chip-14"
                classes={
                  Object {
                    "avatar": "MuiChip-avatar-22",
                    "avatarChildren": "MuiChip-avatarChildren-23",
                    "clickable": "MuiChip-clickable-20",
                    "deletable": "MuiChip-deletable-21",
                    "deleteIcon": "MuiChip-deleteIcon-25",
                    "label": "MuiChip-label-24",
                    "root": "MuiChip-root-19",
                  }
                }
                component="div"
                label="foo"
                onClick={[Function]}
                onDelete={[Function]}
                style={
                  Object {
                    "backgroundColor": undefined,
                    "pointerEvents": undefined,
                  }
                }
              >
                <div
                  className="MuiChip-root-19 MuiChip-clickable-20 MuiChip-deletable-21 ChipInput-chip-14"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="button"
                  style={
                    Object {
                      "backgroundColor": undefined,
                      "pointerEvents": undefined,
                    }
                  }
                  tabIndex={0}
                >
                  <span
                    className="MuiChip-label-24"
                  >
                    foo
                  </span>
                  <pure(Cancel)
                    className="MuiChip-deleteIcon-25"
                    onClick={[Function]}
                  >
                    <Cancel
                      className="MuiChip-deleteIcon-25"
                      onClick={[Function]}
                    >
                      <WithStyles(SvgIcon)
                        className="MuiChip-deleteIcon-25"
                        onClick={[Function]}
                      >
                        <SvgIcon
                          className="MuiChip-deleteIcon-25"
                          classes={
                            Object {
                              "colorAction": "MuiSvgIcon-colorAction-29",
                              "colorDisabled": "MuiSvgIcon-colorDisabled-30",
                              "colorError": "MuiSvgIcon-colorError-31",
                              "colorPrimary": "MuiSvgIcon-colorPrimary-27",
                              "colorSecondary": "MuiSvgIcon-colorSecondary-28",
                              "root": "MuiSvgIcon-root-26",
                            }
                          }
                          color="inherit"
                          onClick={[Function]}
                          viewBox="0 0 24 24"
                        >
                          <svg
                            aria-hidden="true"
                            className="MuiSvgIcon-root-26 MuiChip-deleteIcon-25"
                            focusable="false"
                            onClick={[Function]}
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
                            />
                          </svg>
                        </SvgIcon>
                      </WithStyles(SvgIcon)>
                    </Cancel>
                  </pure(Cancel)>
                </div>
              </Chip>
            </WithStyles(Chip)>
            <WithStyles(Chip)
              className="ChipInput-chip-14"
              key="1"
              label="bar"
              onClick={[Function]}
              onDelete={[Function]}
              style={
                Object {
                  "backgroundColor": undefined,
                  "pointerEvents": undefined,
                }
              }
            >
              <Chip
                className="ChipInput-chip-14"
                classes={
                  Object {
                    "avatar": "MuiChip-avatar-22",
                    "avatarChildren": "MuiChip-avatarChildren-23",
                    "clickable": "MuiChip-clickable-20",
                    "deletable": "MuiChip-deletable-21",
                    "deleteIcon": "MuiChip-deleteIcon-25",
                    "label": "MuiChip-label-24",
                    "root": "MuiChip-root-19",
                  }
                }
                component="div"
                label="bar"
                onClick={[Function]}
                onDelete={[Function]}
                style={
                  Object {
                    "backgroundColor": undefined,
                    "pointerEvents": undefined,
                  }
                }
              >
                <div
                  className="MuiChip-root-19 MuiChip-clickable-20 MuiChip-deletable-21 ChipInput-chip-14"
                  onClick={[Function]}
                  onKeyDown={[Function]}
                  role="button"
                  style={
                    Object {
                      "backgroundColor": undefined,
                      "pointerEvents": undefined,
                    }
                  }
                  tabIndex={0}
                >
                  <span
                    className="MuiChip-label-24"
                  >
                    bar
                  </span>
                  <pure(Cancel)
                    className="MuiChip-deleteIcon-25"
                    onClick={[Function]}
                  >
                    <Cancel
                      className="MuiChip-deleteIcon-25"
                      onClick={[Function]}
                    >
                      <WithStyles(SvgIcon)
                        className="MuiChip-deleteIcon-25"
                        onClick={[Function]}
                      >
                        <SvgIcon
                          className="MuiChip-deleteIcon-25"
                          classes={
                            Object {
                              "colorAction": "MuiSvgIcon-colorAction-29",
                              "colorDisabled": "MuiSvgIcon-colorDisabled-30",
                              "colorError": "MuiSvgIcon-colorError-31",
                              "colorPrimary": "MuiSvgIcon-colorPrimary-27",
                              "colorSecondary": "MuiSvgIcon-colorSecondary-28",
                              "root": "MuiSvgIcon-root-26",
                            }
                          }
                          color="inherit"
                          onClick={[Function]}
                          viewBox="0 0 24 24"
                        >
                          <svg
                            aria-hidden="true"
                            className="MuiSvgIcon-root-26 MuiChip-deleteIcon-25"
                            focusable="false"
                            onClick={[Function]}
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M12 2C6.47 2 2 6.47 2 12s4.47 10 10 10 10-4.47 10-10S17.53 2 12 2zm5 13.59L15.59 17 12 13.41 8.41 17 7 15.59 10.59 12 7 8.41 8.41 7 12 10.59 15.59 7 17 8.41 13.41 12 17 15.59z"
                            />
                          </svg>
                        </SvgIcon>
                      </WithStyles(SvgIcon)>
                    </Cancel>
                  </pure(Cancel)>
                </div>
              </Chip>
            </WithStyles(Chip)>
            <WithStyles(Input)
              classes={
                Object {
                  "input": "ChipInput-input-3",
                  "root": "ChipInput-inputRoot-2",
                }
              }
              disableUnderline={true}
              inputRef={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onKeyPress={[Function]}
              onKeyUp={[Function]}
              placeholder={null}
              value=""
            >
              <Input
                classes={
                  Object {
                    "disabled": "MuiInput-disabled-37",
                    "error": "MuiInput-error-35",
                    "focused": "MuiInput-focused-36",
                    "formControl": "MuiInput-formControl-33",
                    "fullWidth": "MuiInput-fullWidth-40",
                    "inkbar": "MuiInput-inkbar-34",
                    "input": "MuiInput-input-41 ChipInput-input-3",
                    "inputDense": "MuiInput-inputDense-42",
                    "inputDisabled": "MuiInput-inputDisabled-43",
                    "inputMultiline": "MuiInput-inputMultiline-45",
                    "inputSearch": "MuiInput-inputSearch-46",
                    "inputSingleline": "MuiInput-inputSingleline-44",
                    "multiline": "MuiInput-multiline-39",
                    "root": "MuiInput-root-32 ChipInput-inputRoot-2",
                    "underline": "MuiInput-underline-38",
                  }
                }
                disableUnderline={true}
                fullWidth={false}
                inputRef={[Function]}
                multiline={false}
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyPress={[Function]}
                onKeyUp={[Function]}
                placeholder={null}
                type="text"
                value=""
              >
                <div
                  className="MuiInput-root-32 ChipInput-inputRoot-2 MuiInput-formControl-33"
                  onBlur={[Function]}
                  onFocus={[Function]}
                  onKeyPress={[Function]}
                >
                  <input
                    aria-invalid={false}
                    aria-required={false}
                    className="MuiInput-input-41 ChipInput-input-3 MuiInput-inputSingleline-44"
                    disabled={false}
                    onChange={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    placeholder={null}
                    type="text"
                    value=""
                  />
                </div>
              </Input>
            </WithStyles(Input)>
          </div>
        </div>
      </FormControl>
    </WithStyles(FormControl)>
  </ChipInput>
</WithStyles(ChipInput)>
`;
